options/rtld/generic/linker.cpp

```diff
@@ -795,6 +816,163 @@ void ObjectRepository::_parseDynamic(SharedObject *object) {
 	}
 }
 
+void ObjectRepository::_parseVerdef(SharedObject *object) {
+	if(!object->versionDefinitionTableOffset) {
+		if(verbose)
+			mlibc::infoLogger()
+				<< "mlibc: Object " << object->name
+				<< " defines no versions" << frg::endlog;
+		return;
+	}
+
+	if(verbose)
+		mlibc::infoLogger()
+			<< "mlibc: Object " << object->name
+			<< " defines " << object->versionDefinitionCount
+			<< " version(s)" << frg::endlog;
+
+	uintptr_t address =
+		object->baseAddress
+		+ object->versionDefinitionTableOffset;
+
+	for(size_t i = 0; i < object->versionDefinitionCount; i++) {
+		elf_verdef def;
+		memcpy(&def, reinterpret_cast<void *>(address), sizeof(elf_verdef));
+
+		// Required by spec.
+		__ensure(def.vd_version == 1);
+		__ensure(def.vd_cnt >= 1);
+		// TODO(qookie): Handle weak versions.
+		__ensure(!(def.vd_flags & ~VER_FLG_BASE));
+
+		// TODO(qookie): Figure out what to do about VERDEF dependency names.
+		// They don't make much sense to me, and it seems to work if we just ignore them.
+		elf_verdaux aux;
+		memcpy(&aux, reinterpret_cast<void *>(address + def.vd_aux), sizeof(elf_verdaux));
+
+		const char *name =
+			static_cast<const char *>(
+				reinterpret_cast<void *>(
+					object->baseAddress
+					+ object->stringTableOffset + aux.vda_name));
+
+		if(verbose)
+			mlibc::infoLogger()
+				<< "mlibc: Object " << object->name
+				<< " defines version " << name
+				<< " (index " << def.vd_ndx << ")"
+				<< frg::endlog;
+
+		SymbolVersion ver{name, def.vd_hash};
+		object->definedVersions.push(ver);
+		object->knownVersions.insert(def.vd_ndx, ver);
+
+		address += def.vd_next;
+	}
+}
+
+void ObjectRepository::_parseVerneed(SharedObject *object) {
+	if(!object->versionRequirementTableOffset) {
+		if(verbose)
+			mlibc::infoLogger() << "mlibc: Object " << object->name << " requires no versions" << frg::endlog;
+		return;
+	}
+
+	if(verbose)
+		mlibc::infoLogger()
+			<< "mlibc: Object " << object->name
+			<< " requires " << object->versionRequirementCount
+			<< " version(s)" << frg::endlog;
+
+	uintptr_t address =
+		object->baseAddress
+		+ object->versionRequirementTableOffset;
+
+	for(size_t i = 0; i < object->versionRequirementCount; i++) {
+		elf_verneed need;
+		memcpy(&need, reinterpret_cast<void *>(address), sizeof(elf_verneed));
+
+		// Required by spec.
+		__ensure(need.vn_version == 1);
+
+		frg::string_view file =
+			static_cast<const char *>(
+				reinterpret_cast<void *>(
+					object->baseAddress
+					+ object->stringTableOffset + need.vn_file));
+
+		// Figure out the target object from file
+		SharedObject *target = nullptr;
+		for(auto dep : object->dependencies) {
+			if(verbose)
+				mlibc::infoLogger()
+					<< "mlibc: Trying " << dep->name << " (SONAME: "
+					<< dep->soName << ") to satisfy " << file << frg::endlog;
+			if(dep->name == file || (dep->soName && dep->soName == file)) {
```

Comment:
Looking at `findLoadedObject`, we _only_ compare against SONAME. Is the `dep->name == file` needed? Not a big deal but it would be good to understand what the correct behaviour is in both cases

Reply:
glibc matches both the SONAME and the file name of the object (although in the opposite order compared us AFAICU?). Without checking the file name our test case doesn't pass, since it doesn't have a SONAME (glibc loads it fine).